Process profilerOverhead data and serialize it

canaltinova opened the pull request:
Currently we just strip out that information from the profile. We have a deploy preview to visualize that data but since captured profiles don't include that data, it's only possible to see that data if we capture the profile with that deploy preview. After this PR, it's gonna be easier to use the deploy preview.

gregtatum commented on src/profile-logic/process-profile.js:
```diff
@@ -801,6 +803,56 @@ function _processCounters(
+ */
+function _processProfilerOverhead(
+  geckoProfile: GeckoProfile | GeckoSubprocessProfile,
+  // The overhead data is listed independently from the threads, so we need an index that
```
Nice comment 👍

gregtatum commented on src/profile-logic/process-profile.js:
```diff
@@ -801,6 +803,56 @@ function _processCounters(
+  );
+
+  if (!mainThread || !geckoProfilerOverhead) {
+    // Profiler overhad or a main thread weren't found, bail out, and return an empty array.
```
```suggestion
    // Profiler overhead or a main thread weren't found, bail out, and return an empty array.
```

gregtatum replied:
Does the overhead data only exist on the GeckoMain? What happens when we don't record GeckoMain?

canaltinova replied:
Actually we are attaching this to `GeckoMain` because I followed the same implementation of `counters`. I was thinking that counters were doing the correct thing. We are displaying both counters and profilerOverhead under GeckoMain thread and if there is no GeckoMain thread, we don't display them. What do you think? If we want to change this, I think we should change counters as well.

gregtatum commented on src/profile-logic/process-profile.js:
```diff
@@ -801,6 +803,56 @@ function _processCounters(
+  const geckoProfilerOverhead: ?GeckoProfilerOverhead =
+    geckoProfile.profilerOverhead_UNSTABLE;
+  const mainThread = geckoProfile.threads.find(
+    thread => thread.name === 'GeckoMain'
```
There are multiple types of GeckoMain threads, are we sure this is the correct one?

canaltinova replied:
Are there? I thought there is only one `GeckoMain` thread per process. Even if there are, I think getting the first one is  not wrong though.

canaltinova commented on src/types/gecko-profile.js:
```diff
@@ -242,6 +257,7 @@ export type GeckoProfileFullMeta = {|
 
 export type GeckoProfileWithMeta<Meta> = {|
   counters?: GeckoCounter[],
+  profilerOverhead_UNSTABLE?: GeckoProfilerOverhead,
```
I landed the patch that removes `_UNSTABLE` designation on gecko side. It's in Nightly now. Also changed the JSON structure a little bit and put the sample data inside `samples` object. Made the necessary changes here as well.

canaltinova commented on src/profile-logic/process-profile.js:
```diff
@@ -1126,10 +1205,21 @@ export function processProfile(
     updateChannel: geckoProfile.meta.updateChannel,
   };
 
+  const profilerOverhead: ProfilerOverhead[] = nullableProfilerOverhead.reduce(
```
That was my first try but when I do that, `flow` cannot infer the type of `profilerOverhead` correctly and still thinks it's nullable. On the other hand, `reduce` function is much more stable in flow and works without any problem. That's the error I'm getting when I change that line to `filter`:

```js
Error ┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈ src/profile-logic/process-profile.js:1217:48

Cannot assign nullableProfilerOverhead.filter(...) to profilerOverhead because inexact null [1] is incompatible with
exact ProfilerOverhead [2] in array element.

 [1] 1117│   const nullableProfilerOverhead: Array<ProfilerOverhead | null> = [
         :
     1214│   //   },
     1215│   //   []
     1216│   // );
 [2] 1217│   const profilerOverhead: ProfilerOverhead[] = nullableProfilerOverhead.filter(overhead => overhead !== null);
     1218│
     1219│   const result = {
     1220│     meta,

```

canaltinova commented:
@gregtatum I addressed the review comments and added some tests. Also stabilized the `profilerOverhead` in gecko side. So we don't have to worry about there anymore.

gregtatum commented on src/test/fixtures/profiles/gecko-profile.js:
```diff
@@ -711,3 +717,98 @@ export function createGeckoCounter(thread: GeckoThread): GeckoCounter {
+  const profiledDuration =
+    thread.samples.data[thread.samples.data.length - 1][1] -
+    thread.samples.data[0][1];
+  const statistics = {
```
Flow will probably give nicer errors if this has an inline definition. Otherwise it relies on the type checking of the function signature and return value, which can be a bit confusing.

canaltinova replied:
Yeah, you're right. Added it, thanks!

gregtatum commented on src/test/unit/process-profile.test.js:
```diff
@@ -235,6 +236,88 @@ describe('gecko counters processing', function() {
+    );
+  });
+
+  it('offsets the overhead timing for child processes', function() {
```
Thanks for remembering this offsetting step!

canaltinova commented:
> Prooooofiling the prooooofiler.

😄 
Thanks for the review!
